components/autocomplete/index.js

```diff
@@ -19,12 +20,76 @@ import Popover from '../popover';
 
 const { ENTER, ESCAPE, UP, DOWN } = keycodes;
 
+/**
+ * Recursively select the firstChild until hitting a leaf node.
+ * @param {Node} node the node to find the recursive first child.
+ * @returns {Node} the first leaf-node >= node in the ordering.
+ */
+function descendFirst( node ) {
+	let n = node;
+	while ( n.firstChild ) {
+		n = n.firstChild;
+	}
+	return n;
+}
+
+/**
+ * Recursively select the lastChild until hitting a leaf node.
+ * @param {Node} node the node to find the recursive last child.
+ * @returns {Node} the first leaf-node <= node in the ordering.
+ */
+function descendLast( node ) {
+	let n = node;
+	while ( n.lastChild ) {
+		n = n.lastChild;
+	}
+	return n;
+}
+
+/**
+ * Is the node a text node.
+ * @param {?Node} node the node to check.
+ * @returns {boolean} true if the node is a text node.
+ */
+function isTextNode( node ) {
+	return node !== null && node.nodeType === 3;
+}
+
+/**
+ * Return the node only if it is a text node, otherwise return null.
+ * @param {?Node} node the node to filter.
+ * @returns {?Node} the node or null if it is not a text node.
+ */
+function onlyTextNode( node ) {
+	return isTextNode( node ) ? node : null;
+}
+
+/**
+ * Find the index of the last charater in the text that is whitespace.
+ * @param {String} text the text to search.
+ * @param {Number} fromIndex the index to start from.
+ * @returns {Number} the last index of a white space character in the text or -1.
+ */
+function lastIndexOfSpace( text, fromIndex ) {
+	const fromI = fromIndex === undefined ?
+			text.length - 1 :
+			Math.min( text.length - 1, fromIndex );
+	for ( let i = fromI; i >= 0; i-- ) {
+		if ( /\s/.test( text.charAt( i ) ) ) {
```

Comment:
This function is called on every keypress in a paragraph block, which is definitely not ideal, and if we're to keep it, should be subject to optimizing for performance. The RegExp test is not the fastest:

https://jsperf.com/character-whitespace

Reply:
Maybe this could serve better than our own iterating implementation:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/search

Reply:
Well I suppose I could use a regex like:
```javascript
/\s\S*$/.exec(text)
```
Still I have no idea if that is better than a loop searching from the end without benchmarking it and it would probably be highly dependent on the length of the text.

I should point out that:
```javascript
' \t\n\r\v'.indexOf( testChar ) !== -1
```
is NOT equivalent for example it does not match nonbreaking space.

Reply:
I did some benchmarking and even on a bad case the existing code can run 655,000 times per second. I tried the regex that I mention above and of course which is better depends on the length of the string, the regex works better for short strings but once the string becomes long the search from the end wins. In short I don't think there is a good reason to replace this function.